src/radical/pilot/unit_manager.py

```diff
@@ -278,66 +280,74 @@ def _pilot_state_cb(self, pilot, state):
+                    'pilot'   : pilot.uid,
+                    'umgr'    : self.uid,
+                    'control' : {'$in' : ['agent_pending', 'agent']}})
 
-            # update the units to avoid pulling them again next time.
-            # NOTE:  this needs not locking with the unit pulling in the
-            #        _unit_pull_cb, as that will only pull umgr_pending 
-            #        units.
-            uids = [unit['uid'] for unit in units]
+                if not unit_cursor.count():
+                    units = list()
+                else:
+                    units = list(unit_cursor)
 
-            self._session._dbs._c.update({'type'  : 'unit',
-                                          'uid'   : {'$in'     : uids}},
-                                         {'$set'  : {'control' : 'umgr'}},
-                                         multi=True)
-            to_restart = list()
-            for unit in units:
+                self._log.debug("units pulled: %3d (pilot dead)", len(units))
 
-                unit['state'] = rps.FAILED
-                if not unit['description'].get('restartable'):
-                    self._log.debug('unit %s not restartable', unit['uid'])
+                if not units:
                     continue
 
-                self._log.debug('unit %s is  restartable', unit['uid'])
-                unit['restarted'] = True
-                ud = rpcud.ComputeUnitDescription(unit['description'])
-                to_restart.append(ud)
-                # FIXME: increment some restart counter in the description?
-                # FIXME: reference the resulting new uid in the old unit.
-
-            if to_restart and not self._closed:
-                self._log.debug('restart %s units', len(to_restart))
-                restarted = self.submit_units(to_restart)
-                for u in restarted:
-                    self._log.debug('restart unit %s', u.uid)
-
-            # final units are not pushed
-            self.advance(units, publish=True, push=False)
-
-            return True
+                # update the units to avoid pulling them again next time.
+                # NOTE:  this needs not locking with the unit pulling in the
+                #        _unit_pull_cb, as that will only pull umgr_pending
+                #        units.
+                uids = [unit['uid'] for unit in units]
+
+                self._session._dbs._c.update({'type'  : 'unit',
+                                              'uid'   : {'$in'     : uids}},
+                                             {'$set'  : {'control' : 'umgr'}},
+                                             multi=True)
+                to_restart = list()
+                for unit in units:
+
+                    unit['state'] = rps.FAILED
+                    if not unit['description'].get('restartable'):
+                        self._log.debug('unit %s not restartable', unit['uid'])
+                        continue
```

Comment:
Here you want to go to the next iteration of the look or exit the loop completely? If the former wouldn't an `else` statement make sense?

Reply:
Semantically there is no difference, and I am not a fan of nesting, specifically in Python.